arch/src/x86_64/mod.rs

```diff
@@ -394,6 +394,65 @@ pub fn configure_vcpu(
         }
     }
 
+    // Copy the TLB/Cache/Prefetch information (CPUID leaf 0x2)
+    cpuid.retain(|c| c.function != 0x2);
+    let leaf = unsafe { x86_64::__cpuid(0x2) };
+    cpuid
+        .push(CpuIdEntry {
+            function: 0x2,
+            eax: leaf.eax,
+            ebx: leaf.ebx,
+            ecx: leaf.ecx,
+            edx: leaf.edx,
+            ..Default::default()
+        })
+        .map_err(|_| Error::PopulatingCpuid)?;
+    warn!(
+        "[DBG] CPUID leaf [0x2]: eax = {:X}, ebx = {:X}, ecx = {:X}, edx = {:X}",
+        leaf.eax, leaf.ebx, leaf.ecx, leaf.edx
+    );
+
+    // Copy the Deterministic Cache Parameter information (CPUID leaf 0x4)
+    cpuid.retain(|c| c.function != 0x4);
+    let mut subleaf_idx: u32 = 0;
+    loop {
+        // Iterate all sub-leaves for CPUID leave 0x04
+        let leaf = unsafe { std::arch::x86_64::__cpuid_count(0x4, subleaf_idx) };
+
+        // Check if the current sub-leaf contains 'null' Cache Type Field (bits 04-00)
+        if (leaf.eax & 0xf) == 0 {
+            // Set the terminating sub-leaf (w/ 'null' cache type field)
+            cpuid
+                .push(CpuIdEntry {
+                    function: 0x4,
+                    index: subleaf_idx,
+                    eax: 0,
+                    ..Default::default()
+                })
+                .map_err(|_| Error::PopulatingCpuid)?;
+
+            break;
+        }
+
+        cpuid
+            .push(CpuIdEntry {
+                function: 0x4,
+                index: subleaf_idx,
+                // Todo: ignoring the informaiton related to addressable processor IDs
+                eax: leaf.eax & 0xff,
```

Comment:
@rbradford This PR is a draft, as the guest kernel doesn't boot when the CPUID 0x4 is copied to the guest in this PR. I will look into it further and will ask your inputs tomorrow if I can't resolve it.

The guest kernel simply hangs after the kernel prints `[    0.004000] Mountpoint-cache hash table entries: 2048 (order: 2, 16384 bytes)`. The complete log is [here](https://gist.github.com/likebreath/a68b1be7df6968e7e7d4d9f5bd6eabc3). My guess was the bug might be related to the sub-leaf value on `EAX` for bits 14-31 which are related to the addressable processor IDs. That information depend on available vcpus and NUMA nodes for the guest and should not be copied directly from the host. I tried to ignore the values from host and simply set them as 0s, but it does not work either. 

Quotes from the [SDM](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-2a-manual.html) (page 293-294):

```
Bits 25 - 14: Maximum number of addressable IDs for logical processors sharing this cache**, ***.
Bits 31 - 26: Maximum number of addressable IDs for processor cores in the physical 
```

Reply:
You need to populate it with the details from the `CpuTopology` struct.